generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java

```diff
@@ -398,181 +418,365 @@ private void generateClassReader(String className, StructSpec struct,
+                    "newCollection.add(",
+                    String.format(");%n"),
+                    false);
+            } else {
+                buffer.printf("newCollection.add(%s);%n",
+                    primitiveReadExpression(arrayType.elementType()));
+            }
             buffer.decrementIndent();
             buffer.printf("}%n");
+            buffer.printf("%snewCollection%s", assignmentPrefix, assignmentSuffix);
+        } else {
+            throw new RuntimeException("Can't handle variable length type " + type);
         }
+        buffer.decrementIndent();
+        buffer.printf("}%n");
+    }
+
+    private void generateClassFromStruct(String className, StructSpec struct,
+                                         Versions parentVersions) {
+        headerGenerator.addImport(MessageGenerator.STRUCT_CLASS);
+        buffer.printf("@SuppressWarnings(\"unchecked\")%n");
+        buffer.printf("@Override%n");
+        buffer.printf("public void fromStruct(Struct struct, short _version) {%n");
+        buffer.incrementIndent();
+        VersionConditional.forVersions(parentVersions, struct.versions()).
+            allowMembershipCheckAlwaysFalse(false).
+            ifNotMember(__ -> {
+                headerGenerator.addImport(MessageGenerator.UNSUPPORTED_VERSION_EXCEPTION_CLASS);
+                buffer.printf("throw new UnsupportedVersionException(\"Can't read " +
+                    "version \" + _version + \" of %s\");%n", className);
+            }).
+            generate(buffer);
         Versions curVersions = parentVersions.intersect(struct.versions());
-        if (curVersions.empty()) {
-            throw new RuntimeException("Version ranges " + parentVersions +
-                " and " + struct.versions() + " have no versions in common.");
+        if (!messageFlexibleVersions.intersect(struct.versions()).empty()) {
+            buffer.printf("NavigableMap<Integer, Object> _taggedFields = null;%n");
         }
+        buffer.printf("this._unknownTaggedFields = null;%n");
+        VersionConditional.forVersions(messageFlexibleVersions, struct.versions()).
+            ifMember(__ -> {
+                headerGenerator.addImport(MessageGenerator.NAVIGABLE_MAP_CLASS);
+                buffer.printf("_taggedFields = (NavigableMap<Integer, Object>) " +
+                    "struct.get(\"%s\");%n", TAGGED_FIELDS_SECTION_NAME);
+            }).
+            generate(buffer);
         for (FieldSpec field : struct.fields()) {
-            generateFieldFromStruct(field, curVersions);
+            VersionConditional.forVersions(field.versions(), curVersions).
+                alwaysEmitBlockScope(field.type().isArray()).
+                ifNotMember(__ -> {
+                    buffer.printf("this.%s = %s;%n", field.camelCaseName(), fieldDefault(field));
+                }).
+                ifMember(presentVersions -> {
+                    VersionConditional.forVersions(field.taggedVersions(), presentVersions).
```

Comment:
Not a big deal, but there are a few cases where we could use a null check of `_taggedField` instead of a version check. Might make the generated code a little more readable.

Reply:
It would be a bit complex to change now since we're also filtering versions that aren't present at all and so on